### DISCUSSION TITLE
[General] zksyncSingleFactory

### DISCUSSION THREAD

**Discussion by Hedy-chu** [zkSync-Community-Hub/zksync-developers#484] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

By deploying a singleFactory contract that can call the create2 method of contractDeployer to create special contracts

### What is the specific issue or error you're encountering?

After deployment, calling contractDeployer's create2 method is not possible, and it returns 'execution reverted'.

### Can you share the error messages or logs you're receiving, if any?

```sh
reason: 'processing response error',
      code: 'SERVER_ERROR',
      body: '{"jsonrpc":"2.0","error":{"code":3,"message":"execution reverted: Deployment failed","data":"0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000114465706c6f796d656e74206661696c6564000000000000000000000000000000"},"id":49}\n',
      error: [Error],
      requestBody: '{"method":"eth_estimateGas","params":[{"type":"0x2","maxFeePerGas":"0x6553f100","maxPriorityFeePerGas":"0x59682f00","from":"0xf0db7ce565cd7419ec2e6548603845a648f6594f","to":"0xa51baf6a9c0ef5db8c1898d5add92bf3227d6088",
```

### Have you made any recent changes to the contract before encountering this issue?

no

### Are there any external libraries or contracts that your contract interacts with?

yes

### Can you provide the relevant portions of your contract code where the issue is occurring?

```solidity
// SPDX-License-Identifier: MIT

pragma solidity 0.8.20;

import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/SystemContractsCaller.sol";
import { DEPLOYER_SYSTEM_CONTRACT } from "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";

/**
 * @title Singleton Factory
 * @notice Exposes CREATE2 to deploy bytecode on deterministic addresses on zkSync-based network on initialization code and salt.
 * @author Vlad Bochok
 */
contract ZkSyncSingletonFactory {
    function deploy(bytes32 _salt, bytes32 _bytecodehash, bytes calldata _calldata) external payable {
        (bool success, ) = SystemContractsCaller
            .systemCallWithReturndata(
                uint32(gasleft()),
                address(DEPLOYER_SYSTEM_CONTRACT),
                uint128(msg.value),
                abi.encodeCall(
                    DEPLOYER_SYSTEM_CONTRACT.create2,
                    (_salt, _bytecodehash, _calldata)
                )
            );
        require(success, "Deployment failed");
    }
}
```

```typescript
import { Wallet, Provider, utils } from 'zksync-web3'
import * as ethers from 'ethers'

const CREATE2_FACTORY_ADDRESS = '0x52296b11f296E93E802C80C3e2cd3974873d5451'
const CREATE2_FACTORY_ABI = ['function deploy(bytes32 _salt, bytes32 _bytecodehash, bytes calldata _calldata)']

const SALT = '0x18e1e988265bdcd793f3224d203231c44d10b17b9faf570e843210ea3a57e38e'

export default async function deploy(args: any) {
  console.log(`Running deploy script for the permit2 contract`)

  let url: URL
  try {
    url = new URL(args.jsonRpc)
  } catch (error) {
    console.error('Invalid JSON RPC URL', (error as Error).message)
    process.exit(1)
  }

  const wallet = new Wallet(args.privateKey, new Provider({ url: url.href }))

  const hre = require('hardhat')
  const artifact = hre.artifacts.readArtifactSync('Permit2')

  const iface = new ethers.utils.Interface(CREATE2_FACTORY_ABI)
  const calldata = iface.encodeFunctionData('deploy', [SALT, utils.hashBytecode(artifact.bytecode), []])

  const factoryDeps = [artifact.bytecode]

  const tx: ethers.providers.TransactionRequest = {
    to: CREATE2_FACTORY_ADDRESS,

    data: calldata,

    customData: {
      factoryDeps,
    },
  }

  const receipt = await (await wallet.sendTransaction(tx)).wait()
  console.log(`Permit2 deploy transaction hash: ${receipt.transactionHash}`)
}
```

### Have you tried to isolate the problem, and if so, what were the results?

no

### What steps have you already taken to try to resolve the issue?

no

### Additional Details

i I have a question about zksyncSingleFactory（testnet address：0xa51baf6a9c0ef5Db8C1898d5aDD92Bf3227d6088）. Is this address special? Can I deploy a contract with the same code separately and create a new contract through this contract using create2? I tried it on the testnet, but it doesn't work. Could you please explain why?

**Comment by bxpana** [comment]
Thanks for reaching out. Will take some time to look over this later.

**Reply by bxpana** [reply]
@Hedy-chu appreciate your patience. It's taken us a little longer to dig into things, but will have an update for you soon.

**Comment by Raid5594** [comment, marked as answer]
Hi! 

This is not a system contract, hence it can be deployed by anyone. The issue is that the bytecode differs, which affects the behavior.

In general:
1. Update to` zksync-ethers` from `zksync-web3`, as we rely on the former.
2. Make sure `isSystem` flag is set in hardhat config, when interacting with system contracts
3. You could deploy `Create2Factory` for your purposes, which is available in the main branch of `era-contracts`

I redeployed using the same bytecode `ZkSyncSingletonFactory` was deployed with and was able to deploy via new contract.
Also, deploying manually `Create2Factory` and interacting with it works the same way (you have to call create2 vs deploy).

If you have any other questions, please let us now
